File: next_greater.java
//Optimized Approach : Monotonic stack

class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int[] result = new int[nums.length];
        
        if(nums == null || nums.length == 0)
            return result;
        if(nums.length == 1)
            result[0] = -1;
        
        //Fill all elements of result as -1;
        Arrays.fill(result, -1);
        
        //stack contains index nos
        Stack<Integer> stack = new Stack<>();
        
        //circular array so we have to traverse twice so len = 2* n
        for(int i=0; i< 2 * nums.length; i++){
            while(!stack.isEmpty() && nums[i % nums.length] > nums[stack.peek()]){
                int index = stack.pop();
                result[index] = nums[i % nums.length];
            }
            
            if(i < nums.length)
                stack.push(i);
        }
        return result;
    }
}

//Time Complexity :O(n)
//Space Complexity :O(n)

File: Next_greater.py
# Time Complexity :O(N)
# Space Complexity :O(N)
# Did this code successfully run on Leetcode : Yes

# Any problem you faced while coding this :

# Your code here along with comments explaining your approach


from collections import deque
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        st=deque()
        n=len(nums)
        result=[-1 for i in range(n)]
        for i in range(2*n):
            while(len(st) and nums[st[-1]]<nums[i%n]):
                popped=st.pop()
                result[popped]=nums[i%n]
            if(i<n):
                st.append(i)
        return result